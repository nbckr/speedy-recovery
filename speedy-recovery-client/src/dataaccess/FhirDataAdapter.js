// TODO: Many different versions of fields like name, make more resilient (and be careful to work with V3!)
// TODO: Add different mappings for different FHIR versions (as a proof of concept)

// FIXME: Update mapping
// arrow function with parameter

const mapPatientToUser = fhirPatientResource => ({
  //This is a temporary hard-code fix as we have not implemented the searching for a patients' parent
  role:
    fhirPatientResource.id === "f0462936-eb4b-4da1-b45a-fbd96ebf8ccb"
      ? "Parent"
      : fhirPatientResource.resourceType,
  birthDate: fhirPatientResource.birthDate,
  gender: fhirPatientResource.gender,
  name: fhirPatientResource.name[0].family || "Unknown",
  careProvider: fhirPatientResource.careprovider || "Unknown",
  //languagePreferred:"Unknown" || fhirPatientResource.communication[0].language.coding[0].display,
  address:
    fhirPatientResource.address[0].line[0] +
    ", " +
    fhirPatientResource.address[0].city +
    ", " +
    fhirPatientResource.address[0].state +
    ", " +
    fhirPatientResource.address[0].postalCode +
    ", " +
    fhirPatientResource.address[0].country,
  phone:
    fhirPatientResource.telecom.filter(element => element.system === "phone") ||
    [],
  email: fhirPatientResource.email || "Unknown"
});

const mapAppointment = fhirAppResource => ({
  id: fhirAppResource.id,
  title: fhirAppResource.text.div.substring(
    42,
    fhirAppResource.text.div.length - 6
  ),
  status: fhirAppResource.status,
  appType: fhirAppResource.appointmentType.coding[0].display,
  indication: fhirAppResource.indication[0].display,
  priority: fhirAppResource.priority,
  description: fhirAppResource.description,
  supportingInfo: fhirAppResource.supportingInformation[0].reference, //TODO: Fix for multiple supporting info
  start: new Date(fhirAppResource.start),
  end: new Date(fhirAppResource.end),
  created: new Date(fhirAppResource.created),
  comment: fhirAppResource.comment,
  patient: fhirAppResource.participant[0].actor.display,
  practitioner: fhirAppResource.participant[1].actor.display,
  location: fhirAppResource.participant[2].actor.display
});

export { mapPatientToUser, mapAppointment };
