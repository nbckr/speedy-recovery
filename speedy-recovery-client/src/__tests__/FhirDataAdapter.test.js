import {mapAppointmentToCalendar, mapPatientToUser} from "../dataaccess/FhirDataAdapter";
import fhirExamplePatient from "./test_input/fhir_r3/FhirExamplePatient.json";
import fhirShortApp from "./test_input/fhir_r3/FhirShortExampleAppointments.json";

test("map FHIR Patient resource to internal user data model", () => {
  const user = mapPatientToUser(fhirExamplePatient);
  expect(user).toMatchSnapshot();
});

test("map FHIR Appointment resource to calendar data model" , () => {
  const app = mapAppointmentToCalendar(fhirShortApp);
  expect(app).toMatchSnapshot();
});