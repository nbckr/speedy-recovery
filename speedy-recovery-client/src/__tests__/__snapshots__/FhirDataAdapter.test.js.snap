// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map FHIR Appointment resource to calendar data model 1`] = `
Object {
  "appType": "Followup",
  "comment": "Further expand on the results of the MRI and determine the next actions that may be appropriate.",
  "created": 2013-10-10T00:00:00.000Z,
  "description": "Discussion on the results of your recent MRI",
  "end": 2013-12-10T11:00:00.000Z,
  "id": "example",
  "indication": "Severe burn of left ear",
  "location": "South Wing, second floor",
  "patient": "Peter James Chalmers",
  "practitioner": "Dr Adam Careful",
  "priority": 5,
  "start": 2013-12-10T09:00:00.000Z,
  "status": "booked",
  "supportingInfo": "DiagnosticReport/ultrasound",
  "title": "Brian MRI results discussion",
}
`;

exports[`map FHIR Condition resource to internal data model 1`] = `
Object {
  "bodySite": "Left Ear",
  "clinicalStatus": "active",
  "onsetDateTime": 2012-05-24T00:00:00.000Z,
  "severity": "Severe",
  "summary": "Burnt Ear",
  "verificationStatus": "confirmed",
}
`;

exports[`map FHIR Medication resource to internal data model 1`] = `
Object {
  "content": "Vancomycin 500mg powder for infusion solution vial (product)",
  "form": "Injection Solution (qualifier value)",
  "id": "med0301",
  "imageURL": "Vancomycin Image",
  "isBrand": true,
  "isOverTheCounter": false,
  "name": "Vancomycin Hydrochloride (VANCOMYCIN HYDROCHLORIDE)",
  "producer": "Pfizer Laboratories Div Pfizer Inc",
}
`;

exports[`map FHIR Patient resource to internal user data model 1`] = `
Object {
  "address": "534 Erewhon St, PleasantVille, Vic, 3999, US",
  "birthDate": "1974-12-25",
  "careProvider": "Dr Example",
  "email": "peter@live.com",
  "gender": "male",
  "name": "Chalmers",
  "phone": Array [
    Object {
      "system": "phone",
      "use": "work",
      "value": "(03) 5555 6473",
    },
  ],
  "role": "Patient",
}
`;
